docs/howto/system_testing.md

```diff
@@ -106,6 +107,45 @@ volumes:
   mysqldata:
 ```
 
+### Agent service deployer
+
+When using the Agent service deployer, everything works as a Docker Compose service 
+deployer. The only difference is that in this case the `elastic-agent` provided by the stack
+will not be used and is expected that the service referenced in the test config is a custom
+`elastic-agent`. This is useful if you need different capabilities than the provided by the
+`elastic-agent` used by the `elastic-package stack` command. 
+
+```
+version: '2.3'
+services:
+  custom-agent:
+    hostname: custom-agent
+    image: "docker.elastic.co/beats/elastic-agent-complete:8.2.0"
+    pid: host
+    cap_add:
+      - AUDIT_CONTROL
+      - AUDIT_READ
+    user: root
+    healthcheck:
+      test: "elastic-agent status"
+      retries: 180
+      interval: 1s
+    environment:
+      FLEET_ENROLL: "1"
+      FLEET_INSECURE: "1"
+      FLEET_URL: "http://fleet-server:8220"
```

Comment:
Umm, this makes me think that we might need some way to pass variables to these configs. These connection settings depend on the stack and how it is started. For example in #789 I am changing these options, and this would break scenarios with these environment variables.

An option can be to create a different configuration file, expected for example in `_dev/deploy/agent/config.yml`, that is used to "patch" a base agent configuration. So for example in this case, it could be something like this:
```
image: "docker.elastic.co/beats/elastic-agent-complete:8.2.0"
pid: host
cap_add:
  - AUDIT_CONTROL
  - AUDIT_READ
user: root
```
This way every package using this deployer only needs to configure the minimal relevant set of settings. And we can control the settings needed for enrollment, or other settings that may be dependent of `elastic-package stack`.

Another middle-ground option could be to use `env_file` here to include an environment file that would be generated by the deployer, and that includes this kind of environment variables needed by agents to connect with the stack.

Reply:
It looks like you're referring to this [issue](https://github.com/elastic/elastic-package/issues/787#issuecomment-1095148116) (Extend "profiles" with local patches).

Do you think that we need a proposal to sketch the final look and then iterate on that? Maybe we should focus on this specific issue. It might be tricky if we want "patches" to be backward compatible with older stacks.

Reply:
I would consider this deployer a separate thing to possible general patches for profiles or the stack subcommand. I think that something like this "agent" deployer has value on itself. Even if later we consider more advanced configurations for profiles or the stack subcommand.

The problem I see with general local patches for `elastic-package stack` is that they may become a source of many reproducibility problems. You may patch the stack to work on one package, and later you start working on a different package and something unexpectedly doesn't work. Or it may be difficult to know or remember that a certain package needs a patched stack.
This could cause a new set of problems similar to the ones with the packages contained in the registry depending on where `elastic-package stack up` was executed (https://github.com/elastic/elastic-package/pull/599).
And as you mention it may be difficult to support patches with multiple versions of the stack in a general way.

I think it is fine to look for a way to start/patch specialized agents on test time as is being done in this PR. These agents are disposable and developers have more awareness on when they are being started. Packages could also select the version of the agent to use, limiting the problems of supporting multiple versions. Variants could help to test with multiple versions or configurations if there are differences.

And if some day we also have stack-level patches, I think it would be ok if the patches are different to the ones used for this "agent" deployer, at the end they are different things.

Reply:
> I think it is fine to look for a way to start/patch specialized agents on test time as is being done in this PR. 

What about unpatching? Did you plan for this too or is it like the Kubernetes agent, once installed it stays there.

Reply:
> > I think it is fine to look for a way to start/patch specialized agents on test time as is being done in this PR.
> 
> What about unpatching? Did you plan for this too or is it like the Kubernetes agent, once installed it stays there.

No need to unpatch. Current implementation starts this patched agent as a docker compose service, and destroys it on tear down. So it doesn't stay. I think this is a good approach.

Reply:
(If I understand it correctly, please @marc-gr correct me if I am wrong :slightly_smiling_face: )

Reply:
It is correct 👍